R/SEIRmodels.R

```diff
@@ -0,0 +1,749 @@
+      # Get the immunity for susceptible humans and use the complement to modify
+      # the infection rate
+      rate_modifier <- 1 - api$get_variable(human, immunity, susceptible)
+      prob <- pars$infection_rate * rate_modifier
+      individual::fixed_probability_state_change_process(
+        "human", S$name, I$name, prob)
+    }
+    if (pars$includeage) {
+      # Get the age for susceptible humans and use the complement to modify the
+      # infection rate
+      rate_modifier <- 1 - api$get_variable(human, age, susceptible)
+      prob <- pars$age_rate * rate_modifier
+      individual::fixed_probability_state_change_process(
+        "human", S$name, I$name, prob)
+    }
+    if (pars$includelocation) {
+      # Get the location for susceptible humans and use the complement to modify
+      # the infection rate
+      rate_modifier <- 1 - api$get_variable(human, location, susceptible)
+      prob <- pars$location_rate * rate_modifier
+      individual::fixed_probability_state_change_process(
+        "human", S$name, I$name, prob)
+    }
+  }
+}
+
+#' @title Infection_process -> S to I and I2
+#'
+#' @param S susceptible
+#' @param I infected
+#' @param I2 severelyinfected
+#' @param human human
+#' @param immunity immunity
+#' @param age age
+#' @param location location
+#' @param pars parameter list
+#'
+#' @export
+#'
+#' @examples
+#' Individual_S_to_I_and_I2(S, I, I2, human, immunity, age, location, pars)
+#' @importFrom stats runif
+Individual_S_to_I_and_I2 <- function(
+  S, I, I2, human, immunity, age, location, pars = NULL) {
+  function(api) {
+    warnings()
+
+    pars <- Get_parameters_for_sirstochastic(pars)
+    # calculate information for infections, recoveries and births
+    inf <- Infections(
+      length(api$get_state(human, I)) + length(api$get_state(human, I2)),
+      length(api$get_state(human, S)), pars)
+
+    n_to_infect <- inf$n_infections_S
+    susceptible <- api$get_state(human, S)
+
+    if (pars$novariations) {
+      if (length(susceptible) != 0 && length(susceptible) > n_to_infect) {
+        infected <- as.integer(
+          round(susceptible[sample.int(length(susceptible), n_to_infect)]))
+
+        infected1 <- as.integer(pars$infection_rate * infected)
+        severelyinfected <- as.integer(infected - infected1)
+
+        if (length(infected1) != 0 && infected1 != 0) {
+          Validated_state_update(api, human, I, infected1, pars$N)
+        }
+
+        if (length(severelyinfected) != 0 && severelyinfected != 0) {
+          Validated_state_update(api, human, I2, severelyinfected, pars$N)
+        }
+      }
+    }
+    if (pars$includeimmune) {
+      # Get the immunity for susceptible humans and use the complement to modify
+      # the infection rate
+      rate_modifier <- 1 - api$get_variable(human, immunity, susceptible)
+
+      prob <- pars$infection_rate * rate_modifier
+      individual::fixed_probability_state_change_process("human", S, I, prob)
```

Comment:
It seems like `fixed_probability_state_change_process` is not what we want here.

It won't move individuals from S to I, it returns a function which will move individuals from S to I every timestep!

Looking at OJ's issue list, we want to use something like [malariasimulation::bernoulli_multi_p](https://github.com/mrc-ide/malariasimulation/blob/master/R/utils.R#L7).

> Draw infections. This will require using our vector lambda_i. Look at malariasimulation::bernoulli_multi_p for how to draw from a vector of different probabilities

Something like

```suggestion
      api$queue_state_update("human", I, bernoulli_multi_p(length(infected), prob)
```